Use `SmallVec` in `NeighborTable` for cache locality

jakelishman opened the pull request:
### Summary

A reasonable chunk of our time in Sabre is spent reading through the `NeighborTable` to find the candidate swaps for a given layout.  Most coupling maps that we care about have a relatively low number of edges between qubits, yet we needed to redirect to the heap for each individual physical-qubit lookup currently.

This switches from using a `Vec` (which is always a fat pointer to heap memory) to `SmallVec` with an inline buffer space of four qubits. With the qubit type being `u32`, the `SmallVec` now takes up the same stack size as a `Vec` but can store (usually) all the swaps directly inline in the outer `Vec` of qubits.  This means that most lookups of the available swaps are looking in the same (up to relatively small offsets) in memory, which makes the access patterns much easier for prefetching to optimise for.

<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Details and comments

This is a less pronounced effect than either #10782 or #10783, but is still a measurable performance improvement - the example in #10782 went from 2.08(4)s on `main` to 1.96(3)s with this PR (again - the numbers are bigger because Webex is using some of my CPU cycles haha).

jakelishman committed:
Use `SmallVec` in `NeighborTable` for cache locality

A reasonable chunk of our time in Sabre is spent reading through the
`NeighborTable` to find the candidate swaps for a given layout.  Most
coupling maps that we care about have a relatively low number of edges
between qubits, yet we needed to redirect to the heap for each
individual physical-qubit lookup currently.

This switches from using a `Vec` (which is always a fat pointer to heap
memory) to `SmallVec` with an inline buffer space of four qubits.
With the qubit type being `u32`, the `SmallVec` now takes up the same
stack size as a `Vec` but can store (usually) all the swaps directly
inline in the outer `Vec` of qubits.  This means that most lookups of
the available swaps are looking in the same (up to relatively small
offsets) in memory, which makes the access patterns much easier for
prefetching to optimise for.

mtreinish commented on crates/accelerate/src/sabre_swap/neighbor_table.rs:
```diff
@@ -97,10 +103,13 @@ impl NeighborTable {
     }
 
     fn __getstate__(&self) -> Vec<Vec<PhysicalQubit>> {
```
For this the return type doesn't matter too much, it's more about just making sure we return something pyo3 converts to native py types on return so that we can rebuild the state. I'm assuming that there are no smallvec conversions built into pyo3 yet, so we need to do this anyway. If this was at all performance sensitive (which it is not) then I'd suggest just returning a pylist directly to avoid a double iteration, but this is fine.

jakelishman replied:
Oddly, I thought I had done this in terms of `PyList`.  I wonder if I accidentally used a slightly older version of this commit as the PR.  I can change it over without any pain.

jakelishman replied:
Done in 5f313e22d.

mtreinish reviewed (dismissed):
LGTM this is a very straighforward change and moving the data structure to small vec makes a lot of sense. I have one small question inline but it's non blocking more just curiosity than anything to change.

jakelishman committed:
Pickle via `PyList` instead of duplicate conversion

`SmallVec` doesn't have implementations of the PyO3 conversion trait, so
it needs to be done manually.  The previous state used to convert to a
Rust-space `Vec` that then needed to have its data moved from the Python
heap to the Rust heap.  This instead changes the conversions to interact
directly with Python lists, rather than using intermediary structures.
